### PR TITLE
UML-396: Lists organisations who viewed lpa using access code

### PR DESCRIPTION
**Pull request by SeemaMenon**
# Purpose

https://opgtransform.atlassian.net/browse/UML-396

Fixes UML-396

## Approach

_Explain how your code addresses the purpose of the change_

## Learning

_Any tips and tricks, blog posts or tools which helped you. Plus anything notable you've discovered about the Use a Lasting Power of Attorney service_

## Checklist

* [X] I have performed a self-review of my own code
* [ ] I have added relevant logging with appropriate levels to my code
* [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [X] I have added tests to prove my work
* [ ] I have added welsh translation tags and updated translation files
* [ ] The product team have tested these changes

### CODE REVIEW

File: service-api/app/features/actor-check-access-codes.feature [hunk 1]
```diff
@@ -28,3 +28,11 @@ Feature: The user is able to check the access codes they have created
     And Co-actors have also created access codes for the same LPA
     When I click to check the access codes
     Then I can see all of the access codes and their details
+
+  @acceptance
+  Scenario: As a user I can see the access codes I have created
```
Comment (GemTay) [inline service-api/app/features/actor-check-access-codes.feature:33]: The scenario could be more descriptive eg "As a user I can see which organisations have used access codes to view an LPA"

File: service-api/app/features/context/Acceptance/LpaContext.php [hunk 2]
```diff
@@ -1585,4 +1587,144 @@ public function iCanSeeAllOfTheAccessCodesAndTheirDetails()
...
+            'Expires' => '2021-01-01T00:00:00Z',
+            'UserLpaActor' => $this->userLpaActorToken,
+            'Organisation' => $this->organisation,
+            'ViewerCode' => $this->accessCode,
+            'Viewed' => [
+                0 => [
+                    'Viewed' => '2020-10-01T00:00:00Z',
+                    'ViewerCode' => $this->accessCode,
+                    'ViewedBy' => 'Test',
+                ],
```
Comment (GemTay) [inline service-api/app/features/context/Acceptance/LpaContext.php:1624]: Maybe repeat the above 'Viewed' array to test that a list of orgs are returned who have viewed the LPA using the same code

File: service-api/app/test/AppTest/DataAccess/DynamoDb/ViewerCodeActivityTest.php [hunk 3]
```diff
@@ -86,7 +86,11 @@ public function viewerCodesStatusSetToTrueWithQueryMatch()
                             'S' => $testCodes[0]['ViewerCode']
                         ],
                         'Viewed' => [
-                            'B' => true,
+                            0 => [
```
Comment (GemTay) [inline service-api/app/test/AppTest/DataAccess/DynamoDb/ViewerCodeActivityTest.php:89]: Again, best to test with more than one here as a list of orgs who have viewed using the same code is the feature

File: service-front/app/features/actor-check-access-codes.feature [hunk 4]
```diff
@@ -46,4 +46,20 @@ Feature: The user is able to check the access codes they have created
     Examples:
       | status    |
       | CANCELLED   |
+
+  @ui
```
Comment (GemTay) [inline service-front/app/features/actor-check-access-codes.feature:50]: Integration tests too

File: service-api/app/features/actor-check-access-codes.feature [hunk 5]
```diff
@@ -28,3 +28,11 @@ Feature: The user is able to check the access codes they have created
     And Co-actors have also created access codes for the same LPA
     When I click to check the access codes
     Then I can see all of the access codes and their details
+
+  @acceptance
```
Comment (GemTay) [inline service-api/app/features/actor-check-access-codes.feature:32]: Integration tests needed too

**Review by GemTay — CHANGES_REQUESTED** [review]
Few changes to be made :)

File: service-api/app/test/AppTest/DataAccess/DynamoDb/ViewerCodeActivityTest.php [hunk 6]
```diff
@@ -97,9 +115,21 @@ public function viewerCodesStatusSetToTrueWithQueryMatch()
...
+        }
+        $this->assertEquals('Some organisation1', $result[0]['Viewed'][0]['ViewedBy']);
+        $this->assertEquals('Some organisation2', $result[0]['Viewed'][1]['ViewedBy']);
+
+//        $this->assertEquals($viewerCode, $result[0]['ViewerCode']);
+//        $this->assertNotEmpty($result[0]['Viewed'][0]);
+//        $this->assertEquals($viewerCode, $result[0]['Viewed'][0]['ViewerCode']);
+//        $this->assertNotEmpty($result[0]['Viewed'][0]['Viewed']);
+//        $this->assertEquals($viewerCode, $result[0]['Viewed'][1]['ViewerCode']);
+//        $this->assertNotEmpty($result[0]['Viewed'][1]['Viewed']);
```
Comment (GemTay) [inline service-api/app/test/AppTest/DataAccess/DynamoDb/ViewerCodeActivityTest.php:132]: Comments to be removed

File: service-api/app/features/context/Acceptance/LpaContext.php (hunk 2 above)
Comment (GemTay) [reply service-api/app/features/context/Acceptance/LpaContext.php:1624]: Just need to repeat the array here again

File: service-api/app/features/context/Acceptance/LpaContext.php [hunk 7]
```diff
@@ -1585,4 +1587,144 @@ public function iCanSeeAllOfTheAccessCodesAndTheirDetails()
...
+            '/v1/lpas/' . $this->userLpaActorToken . '/codes',
+            [
+                'user-token' => $this->userId,
+            ]
+        );
+
+        $this->ui->assertSession()->statusCodeEquals(StatusCodeInterface::STATUS_OK);
+        $response = $this->getResponseAsJson();
+
+        assertArrayHasKey('Viewed', $response[0]);
```
Comment (GemTay) [inline service-api/app/features/context/Acceptance/LpaContext.php:1728]: add some more assertions here to check the "Viewed" array contains the organisations who viewed the LPA using the code

File: service-front/app/features/context/Integration/AccountContext.php [hunk 8]
```diff
@@ -1850,4 +1850,27 @@ public function iAccessTheAccountCreationPage()
...
+    public function iCanSeeTheNameOfTheOrganisationThatViewedTheLPA()
+    {
+        // Not needed for this context
+    }
+
+    /**
+     * @When /^I click to check my access codes that is used to view LPA/
+     */
+    public function iClickToCheckMyAccessCodeThatIsUsedToViewLPA()
+    {
```
Comment (GemTay) [inline service-front/app/features/context/Integration/AccountContext.php:1874]: this behat step needs filling in. You can probably copy the existing iClickToCheckMyAccessCodes integration step and change the "Viewed" data in the api response. Then update the assertions to check the correct data was returned

**Review by GemTay — CHANGES_REQUESTED** [review]
a few more tweaks :)

File: service-front/app/features/context/Integration/AccountContext.php (hunk 8 above)
Comment (SeemaMenon) [reply service-front/app/features/context/Integration/AccountContext.php:1874]: This was a clear miss .. :(. Added in now.

File: service-api/app/test/AppTest/DataAccess/DynamoDb/ViewerCodeActivityTest.php (hunk 6 above)
Comment (SeemaMenon) [reply service-api/app/test/AppTest/DataAccess/DynamoDb/ViewerCodeActivityTest.php:132]: My mistake, sorry

**Review by GemTay — APPROVED** [review]
Nice job 👍

File: service-api/app/features/context/Acceptance/LpaContext.php [hunk 9]
```diff
@@ -1585,4 +1587,278 @@ public function iCanSeeAllOfTheAccessCodesAndTheirDetails()
...
+    {
+        // Not needed for this context
+    }
+
+    /**
+     * @When /^I click to check my access codes that is used to view LPA$/
+     */
+    public function iClickToCheckMyAccessCodesThatIsUsedToViewLPA()
+    {
+//        $code1 = [
```
Comment (GemTay) [inline service-api/app/features/context/Acceptance/LpaContext.php:1612]: Commented code

**Review by GemTay — CHANGES_REQUESTED** [review]
Comments to be removed

File: service-api/app/features/context/Acceptance/LpaContext.php (hunk 9 above)
Comment (SeemaMenon) [reply service-api/app/features/context/Acceptance/LpaContext.php]: That's unforgivable !!!! Don't kill me Gemma... 😿 

Removed now.

**Review by GemTay — APPROVED** [review]

**Pull request merged**